Fix #3655: Allow dots (.), hyphens (-), and underscores (_) in GitHub usernames

Nisarg908 opened the pull request:
🔧 What does this PR do?
This PR fixes username validation in the modules create flow. Previously, usernames with dots (e.g., @john.doe) were rejected as invalid, even though they are allowed on platforms like GitLab. This blocked users from using custom module repositories with valid usernames.

🛠️ Changes made:
Updated the GitHub username validation regex in create.py to allow ., -, and _ characters

Error message updated to reflect valid formats

📌 Why is this needed?
To support usernames on platforms other than GitHub (e.g., GitLab), which allow characters like . in usernames. Previously, the validation was too strict and caused confusion and unnecessary errors for users.

🔗 Related Issue
Fixes #3655 

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] `CHANGELOG.md` is updated
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Documentation in `docs` is updated

mashehu commented:
Nice, thanks!

Could you add a check that we are not working with a github repo and then add a pytest for the whole new functionality?

Nisarg908 commented:
Hi @mashehu 
Just to check I’ve got the requirement right: when the project’s remote isn’t on github.com, you’d like us to skip the GitHub‑specific username regex and only require the username to start with “@”, correct?

mashehu commented:
Not too sure about the requirements for gitlab, but maybe this would be the simplest solution

Nisarg908 commented:
Hi @mashehu,

I've implemented the new logic for username validation, including conditional checks to bypass strict regex validation for non-GitHub origins. I’ve also added corresponding unit tests to ensure correctness, and all test cases are passing now.

Could you please review the changes when you get a chance? Let me know if any modifications or improvements are needed.

mashehu commented on nf_core/components/create.py:
```diff
@@ -418,16 +418,41 @@ def _get_username(self):
             log.debug(f"Could not find GitHub username using 'gh' cli command: [red]{e}")
+        # Determine whether this repo is hosted on GitHub
+        repo_url = (
+            self.wf.config.get("manifest.profile_repository")
```
this config entry doesn't exist. you would need to check it based on modules_repo_url or remote_url

Nisarg908 replied:
Done Please review

mashehu commented on tests/test_create_username.py:
```diff
@@ -0,0 +1,52 @@
+import pytest
```
sorry, I didn't point you to the right location before. the tests should be part of https://github.com/nf-core/tools/blob/main/tests/modules/test_create.py

Nisarg908 replied:
Added the test function in test_create.py

mashehu commented:
Hi, yes, I will come back to this on Monday. Sorry for the delay, but I am on partial parental leave and we had a release last week that took away a lot of time.

mashehu commented on CHANGELOG.md:
```diff
@@ -35,6 +35,7 @@
 ### Modules
 
 - Remove args stub from module template to satisfy language server ([#3403](https://github.com/nf-core/tools/pull/3403))
+- Allow GitHub usernames with dots (`.`), hyphens (`-`), and underscores (`_`) in `nf-core modules create` to support usernames from GitLab and other providers. ([#3658](https://github.com/nf-core/tools/pull/3658))
```
```suggestion
- Allow usernames with dots (`.`), hyphens (`-`), and underscores (`_`) in `nf-core modules create` to support usernames from GitLab and other providers. ([#3658](https://github.com/nf-core/tools/pull/3658))
```

mashehu commented on nf_core/components/create.py:
```diff
@@ -430,16 +430,41 @@ def _get_username(self):
+        # Determine whether this repo is hosted on GitHub
+        repo_url = (
+            getattr(self.modules_repo, "remote_url", None)
+            or getattr(self.wf.dir, "git_origin_url", "")
```
i don't thing `wf` is an attribute for `ComponentCreate` instances

mashehu commented on tests/modules/test_create.py:
```diff
@@ -164,6 +164,49 @@ def test_modules_migrate_symlink(self, mock_rich_ask):
+    def test_username_validation(self, origin_url, username, should_pass, monkeypatch):
+        """Test GitHub username validation in _get_username logic"""
+
+        class DummyWF:
+            def __init__(self, origin_url):
+                self.config = {}
+                self.dir = type("Dir", (), {"git_origin_url": origin_url})
+
+        class FakeModulesRepo:
+            no_pull_global = False
+            def __init__(self, *args, **kwargs):
+                pass
```
see `tests/test_modules.py` on how we initiate test module repos
